### PR TITLE
Calo DB Tables and Associated Files

### PR DESCRIPTION
**Pull request by sophiemiddleton**
First set of tables for calorimeter calibration and passing of the constants to the Offline code. Follow up PRs will come once this is in.

### CODE REVIEW

**Comment by sophiemiddleton** [comment]
@FNALbuild run build test

**Comment by sophiemiddleton** [comment]
oh, could do. I'm not sure, its just tables at this point, they arent been used in reco explicitly

**Comment by sophiemiddleton** [comment]
@FNALbuild run build test

**Comment by sophiemiddleton** [comment]
any update on this PR?

**Comment by brownd1978** [comment]
@FNALbuild run build test

File: DbTables/src/DbTableFactory.cc [hunk 1]
```diff
@@ -57,6 +67,8 @@ mu2e::DbTable::ptr_t mu2e::DbTableFactory::newTable(std::string const& name) {
     return std::shared_ptr<mu2e::DbTable>(new mu2e::SimEfficiencies2());
   } else if (name == "CalRoIDMapDIRACToOffline") {
     return std::shared_ptr<mu2e::DbTable>(new mu2e::CalRoIDMapDIRACToOffline());
+  } else if (name=="CalRoIDMapOfflineToDIRAC") {
+    return std::shared_ptr<mu2e::DbTable>(new mu2e::CalRoIDMapOfflineToDIRAC());
```
Comment (rlcee) [inline DbTables/src/DbTableFactory.cc:71]: adding this duplicate line looks like a mistake

File: CaloConditions/src/CalCalibMaker.cc [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+/*
+Written by S. Middleton
+*/
+namespace mu2e {
+  typedef std::shared_ptr<CalCalib> ptr_t;
+
+  ptr_t CalCalibMaker::fromFcl() {
+  CalCalibPar nominal(_config.ADC2MeV(), _config.timeoffset(), _config.algName());
+
+  size_t nChan = 1; //FIXME
```
Comment (rlcee) [inline CaloConditions/src/CalCalibMaker.cc:13]: can become CaloConst::_nChannel

File: CaloConditions/src/CalCalibMaker.cc [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  CalCalib::CalibVec cvec(nChan, nominal);
+  auto ptr = std::make_shared<CalCalib>(cvec);
+  return ptr;
+
+  } // end fromFcl
+
+  ptr_t CalCalibMaker::fromDb(CalEnergyCalib::cptr_t) {
+    // initially fill from fcl to get all the constants
+
+    auto ptr = fromFcl();
```
Comment (rlcee) [inline CaloConditions/src/CalCalibMaker.cc:24]: I expected this to contain code to fill CalCalib from the db table argument. It is fine if you want to write that code later.

File: CaloConditions/src/CalCalib.cc [hunk 4]
```diff
@@ -0,0 +1 @@
+#include "Offline/CaloConditions/inc/CalCalib.hh"
```
Comment (rlcee) [inline CaloConditions/src/CalCalib.cc:1]: These entities usually have dump print triggered by the verbose flag.

File: CaloConditions/inc/CalCalibCache.hh [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+#include "Offline/Mu2eInterfaces/inc/ProditionsCache.hh"
+#include "Offline/CaloConditions/inc/CalCalibMaker.hh"
+#include "Offline/ProditionsService/inc/ProditionsHandle.hh"
+
+
+namespace mu2e {
+  class CalCalibCache : public ProditionsCache {
+  public: 
+    CalCalibCache(CalCalibConfig const& config):
+      ProditionsCache(CalEnergyCalib::cxname,config.verbose()),
```
Comment (rlcee) [inline CaloConditions/inc/CalCalibCache.hh:17]: Should be CalCalib::cxname

File: CaloConditions/inc/CalCalibCache.hh [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    void initialize() {
+     if(_useDb) {
+        _calenergycalib_p = std::make_unique<DbHandle<CalEnergyCalib>>();
+      }
+    }
+    
+    set_t makeSet(art::EventID const& eid) {
+      ProditionsEntity::set_t cids;
+      if(_useDb) {  
+        auto cal = _calenergycalib_p->get(eid);
```
Comment (rlcee) [inline CaloConditions/inc/CalCalibCache.hh:29]: varible cal unusued, could leave it out

File: CaloConditions/inc/CalCalibCache.hh [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        iov.overlap(_calenergycalib_p->iov());
+      }
+      return iov;
+    }
+    
+    ProditionsEntity::ptr makeEntity(art::EventID const& eid) {
+       if(_useDb) {
+        return _maker.fromDb( _calenergycalib_p->getPtr(eid)); 
+      } else {
+	      return _maker.fromFcl();
```
Comment (rlcee) [inline CaloConditions/inc/CalCalibCache.hh:49]: indent typo

File: DbTables/inc/CalCosmicEnergyCalib.hh [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+      }
+   _rows.emplace_back(CaloSiPMId(index),std::stof(columns[1]),std::stof(columns[2]),std::stof(columns[3]),std::stof(columns[4]),std::stof(columns[5]));
+
+  }
+
+    void rowToCsv(std::ostringstream& sstream, std::size_t irow) const override {
+      Row const& r = _rows.at(irow);
+      sstream << std::fixed << std::setprecision(5);
+      sstream << 0 <<","; //TODO
```
Comment (rlcee) [inline DbTables/inc/CalCosmicEnergyCalib.hh:62]: r.roid()

File: DbTables/inc/CalEnergyCalib.hh [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    std::size_t nrow() const override { return _rows.size(); };
+    size_t size() const override { return baseSize() + nrow()*nrow()/2 + nrow()*sizeof(Row); };
+
+    void addRow(const std::vector<std::string>& columns) override {
+      std::uint16_t index = std::stoul(columns[0]);
+    // enforce order, so channels can be looked up by index
+    if (index!=int(_rows.size())) {
+        throw cet::exception("CALOENERGYCALIB_BAD_INDEX")<<"CalEnergyCalib::addRow found index out of order:"<<index << " != " << int(_rows.size()) <<"\n";
+      }
+       _rows.emplace_back(CaloSiPMId(index),std::stof(columns[1]),std::stof(columns[2]),std::stof(columns[3]));
```
Comment (rlcee) [inline DbTables/inc/CalEnergyCalib.hh:58]: last one is stoi

File: DbTables/inc/CalCosmicTimeCalib.hh [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+      }
+
+ _rows.emplace_back(CaloSiPMId (index),std::stof(columns[1]),std::stof(columns[2]),std::stof(columns[3]),std::stof(columns[4]),std::stof(columns[5]));
+
+    }
+
+    void rowToCsv(std::ostringstream& sstream, std::size_t irow) const override {
+      Row const& r = _rows.at(irow);
+      sstream << std::fixed << std::setprecision(5);
+      sstream << 0 <<","; //TODO
```
Comment (rlcee) [inline DbTables/inc/CalCosmicTimeCalib.hh:62]: r.roid()

**Review by rlcee — CHANGES_REQUESTED** [review]
- The word "calib" in the database tables is a bit redundant (but your choice).
- "algName" is misleading since it is an int, maybe should be called "algId" and come from an enum (if its values are small and fixed)
- I would remove the "rowAt" accessor for all tables because it is redundant. It is relevant if you had maps instead of just a dense vector.
- I would make the argument of the row accessor an actual CalSiPMId
- please remove the term " nrow()*nrow()/2" from the size accessor, that accounts for the map, which you don't have
- please add two accessors for the tables:
```
  virtual std::size_t nrowFix() const override { return CaloConst::_nChannel; };
  const std::string orderBy() const { return std::string("roid"); }

```

File: DbTables/src/DbTableFactory.cc (hunk 1 above)
Comment (sophiemiddleton) [reply DbTables/src/DbTableFactory.cc:71]: these are different tables, right?

File: CaloConditions/src/CalCalibMaker.cc (hunk 3 above)
Comment (sophiemiddleton) [reply CaloConditions/src/CalCalibMaker.cc:24]: yes it will be part of the next iteration

File: CaloConditions/inc/CalCalibCache.hh (hunk 5 above)
Comment (sophiemiddleton) [reply CaloConditions/inc/CalCalibCache.hh:17]: done

File: CaloConditions/inc/CalCalibCache.hh (hunk 7 above)
Comment (sophiemiddleton) [reply CaloConditions/inc/CalCalibCache.hh:49]: fixed

File: DbTables/inc/CalCosmicEnergyCalib.hh (hunk 8 above)
Comment (sophiemiddleton) [reply DbTables/inc/CalCosmicEnergyCalib.hh:62]: fixed

File: DbTables/inc/CalEnergyCalib.hh (hunk 9 above)
Comment (sophiemiddleton) [reply DbTables/inc/CalEnergyCalib.hh:59]: fixed

File: DbTables/src/DbTableFactory.cc (hunk 1 above)
Comment (rlcee) [reply DbTables/src/DbTableFactory.cc:71]: I think you accidentally copied two lines related to CalRoIDMapOfflineToDIRAC, so they are duplicates, the lines showing up green here

File: CaloConditions/inc/CalCalib.hh [hunk 11]
```diff
@@ -30,10 +30,13 @@ namespace mu2e {
       const CalCalibPar& calib(std::uint16_t roid) const {
         return _cvec.at(roid);
       }
+      CalCalibPar const* calCalib() const { return _calpar.get(); }
+      void print( std::ostream& ) const;
 
   private:
 
       CalibVec _cvec;
+      std::unique_ptr<CalCalibPar> _calpar;
```
Comment (rlcee) [inline CaloConditions/inc/CalCalib.hh:39]: I don't understand this.  I expect the entity to hold a CalCalibPar for each channel which is there as _cvec, so why is this lone one here?  The print dump should loop over _cvec.

Comment (sophiemiddleton) [reply CaloConditions/inc/CalCalib.hh:39]: oops, I was trying to figure something out but should remove now

**Review by brownd1978 — COMMENTED** [review]
I have one general question about this PR: the classes use the prefex 'Cal', whereas all the rest of the calorimeter code uses 'Calo'.   Is this difference intentional?

**Comment by rlcee** [comment]
> I have one general question about this PR: the classes use the prefex 'Cal', whereas all the rest of the calorimeter code uses 'Calo'. Is this difference intentional?

The database code requires regular patterns: c++ classes like XyzDef, which correspond to postgres schema xyz and table name def within that schema.  I made the schemas trk, cal, crv, stm, tst, val long ago.  It would not be wildly inconvenient at this time to change the cal schema to calo.  It will be ~impossible in the future.  Technically, is is also possible to break the correspondence between c++ and schema.table, but that is just inviting confusion.

**Review by brownd1978 — APPROVED** [review]
Thanks for the explanation of the name

File: CaloConditions/src/CalCalib.cc [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+#include <sstream>
+
+using namespace std;
+
+namespace mu2e {
+
+  void CalCalib::print( ostream& out) const{
+    out << "ADCMeV "
+      <<  ( _calpar->ADC2MeV() ) << endl;
```
Comment (rlcee) [inline CaloConditions/src/CalCalib.cc:11]: This reference also need to be fixed, thanks

Comment (sophiemiddleton) [reply CaloConditions/src/CalCalib.cc]: OK, I've removed this file, I dont think its needed yet, if at all

**Review by rlcee — APPROVED** [review]

**Comment by brownd1978** [comment]
@FNALbuild run build test

**Comment by sophiemiddleton** [comment]
Any further comments on this PR?

**Comment by brownd1978** [comment]
Note that the merge of 989 uncovered a minor conflict, please resolve and

**Comment by sophiemiddleton** [comment]
@FNALbuild run build test

**Pull request merged**